pkg/ads/ads.go

```diff
@@ -14,154 +14,146 @@
-		vips := convertToVIPs(addresses)
-		if len(vips) == 0 {
-			continue
-		}
-		for _, host := range e.Hosts {
-			key := fmt.Sprintf("%s.", host)
-			if strings.Contains(host, "*") {
-				// Validation will ensure that the host is of the form *.foo.com
-				parts := strings.SplitN(host, ".", 2)
-				// Prefix wildcards with a . so that we can distinguish these entries in the map
-				key = fmt.Sprintf(".%s.", parts[1])
-			}
-			dnsEntries[key] = vips
-		}
-	}
-	c.server.UpdateDNSEntries(dnsEntries)
-
-	// send ACK
-	ackRequest := &discovery.DiscoveryRequest{
-		VersionInfo:   response.VersionInfo,
-		ResponseNonce: response.Nonce,
-		TypeUrl:       response.TypeUrl,
-	}
-	if err := con.Send(ackRequest); err != nil {
-		fmt.Println("Error sending ACK:", err)
-	}
-}
-
-func convertToVIPs(addresses []string) []net.IP {
-	vips := make([]net.IP, 0)
-
-	for _, address := range addresses {
-		// check if its CIDR.  If so, reject the address unless its /32 CIDR
-		if strings.Contains(address, "/") {
-			if ip, network, err := net.ParseCIDR(address); err != nil {
-				ones, bits := network.Mask.Size()
-				if ones == bits {
-					// its a full mask (e.g., /32). Effectively an IP
-					vips = append(vips, ip)
-				}
-			}
-		} else {
-			if ip := net.ParseIP(address); ip != nil {
-				vips = append(vips, ip)
-			}
-		}
-	}
-
-	return vips
-}
+ import (
+	 "fmt"
+	 "log"
+	 "net"
+	 "strings"
+ 
+	 discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	 "google.golang.org/grpc"
+	 "google.golang.org/grpc/credentials/insecure"
+	 "google.golang.org/protobuf/proto"
+	 "google.golang.org/protobuf/types/known/anypb"
+	 mcp "istio.io/api/mcp/v1alpha1"
+	 networking "istio.io/api/networking/v1alpha3"
+	 "istio.io/istio/pkg/adsc"
+	 "istio.io/istio/pkg/config/schema/collections"
+	 "kmesh.net/kmesh-coredns-plugin/pkg/dns"
+	 "kmesh.net/kmesh-coredns-plugin/pkg/options"
+ )
+ 
+ type AdsController struct {
+	 server    *dns.KmeshDNSServer
+	 adsClient *adsc.ADSC
+	 vip       string
+ }
+ 
+ func NewAdsController(dnsServer *dns.KmeshDNSServer) (*AdsController, error) {
+	 c := &AdsController{
+		 server: dnsServer,
+		 vip:    options.GetConfig().VIP,
+	 }
+	 client, err := adsc.New(options.GetConfig().XDSAddress, &adsc.ADSConfig{
+		 InitialDiscoveryRequests: configInitialRequests(),
+		 ResponseHandler:          c,
+		 Config: adsc.Config{
+			 Workload:  options.GetConfig().ServiceNode,
+			 Namespace: options.GetConfig().ServiceNameSpace,
+			 GrpcOpts: []grpc.DialOption{
+				 // Because we use the custom grpc options for adsc, here we should
+				 // explicitly set transport credentials.
+				 // TODO: maybe we should use the tls settings within ConfigSource
+				 // to secure the connection between istiod and remote xds server.
+				 grpc.WithTransportCredentials(insecure.NewCredentials()),
+			 },
+		 },
+	 })
+	 if err != nil {
+		 log.Fatalf("failed to dial XDS %s %v", "controller.gateway-system.svc.cluster.local:15010", err)
```

Comment:
prevent hardcoding, and i think controller.gateway-system.svc.cluster.local:15010 is not right